File: slidingWindow/76_minimum_window_substring.py
class Solution:
    # O(m + n) time | O(m + n) space - where m is s.length and n is t.length
    def minWindow(self, s: str, t: str) -> str:
        countT, window = Counter(t), {}
        have, need = 0, len(countT)
        res, resLen = [-1, -1], float('inf')
        l = 0
        for r in range(len(s)):
            c = s[r]
            window[c] = window.get(c, 0) + 1
            
            if c in countT and window[c] == countT[c]:
                have += 1
            
            while have == need:
                if (r - l + 1) < resLen:
                    res = [l, r]
                    resLen = r - l + 1
                
                window[s[l]] -= 1
                if s[l] in countT and window[s[l]] < countT[s[l]]:
                    have -= 1
                l += 1
        
        l, r = res
        return s[l:r+1] if resLen != float('inf') else ""

File: slidingWindow/76_minimum_window_substring.go
func minWindow(s string, t string) string {
    countT, window := make(map[byte]int), make(map[byte]int)
    for i := range t {
        countT[t[i]]++
    }
    have, need := 0, len(countT)
    res, resLen := []int{-1, -1}, math.Inf(1)
    
    l := 0
    for r := range s {
        window[s[r]]++
        if count, exist := countT[s[r]]; exist && window[s[r]] == count {
            have++
        }
        
        for have == need {
            if float64(r - l + 1) < resLen {
                resLen = float64(r - l + 1)
                res = []int{l, r}
            }
            window[s[l]]--
            if count, exist := countT[s[l]]; exist && window[s[l]] < count {
                have--           
            }
            l++
        }
    }
    if float64(resLen) != math.Inf(1) {
        return s[res[0]:res[1]+1]
    }
    return ""
}
